tests/integration/test_legacy_mysql.py

```diff
@@ -0,0 +1,208 @@
+
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+CLUSTER_NAME = "test_cluster"
+DATABASE_APP_NAME = "mysql"
+OSM_KEYSTONE_APP_NAME = "osm-keystone"
+KFP_API_APP_NAME = "kfp-api"
+
+
+@pytest.mark.order(1)
+@pytest.mark.skip_if_deployed
+@pytest.mark.legacy_mysql_tests
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy the mysql charm.
+
+    Args:
+        ops_test: The ops test framework
+    """
+    async with ops_test.fast_forward():
+        # Build and deploy the mysql charm
+        charm = await ops_test.build_charm(".")
+        resources = {"mysql-image": METADATA["resources"]["mysql-image"]["upstream-source"]}
+        config = {
+            "mysql-interface-user": "keystone",
+            "mysql-interface-database": "keystone",
+        }
+        await ops_test.model.deploy(
+            charm,
+            resources=resources,
+            config=config,
+            application_name=DATABASE_APP_NAME,
+            num_units=3,
+        )
+
+        await ops_test.model.wait_for_idle(
+            apps=[DATABASE_APP_NAME],
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+            wait_for_exact_units=3,
+        )
+        assert len(ops_test.model.applications[DATABASE_APP_NAME].units) == 3
+        for unit in ops_test.model.applications[DATABASE_APP_NAME].units:
+            assert unit.workload_status == "active"
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.legacy_mysql_tests
+async def test_osm_keystone_bundle_mysql(ops_test: OpsTest) -> None:
+    """Deploy the osm keystone bundle to test the legacy 'mysql' relation.
+
+    Args:
+        ops_test: The ops test framework
+    """
+    async with ops_test.fast_forward():
+        # Build and deploy the osm-keystone charm
+        # (using ops_test.juju instead of ops_test.deploy as the latter does
+        # not correctly deploy with the correct resources)
+        # TODO: Replace edge channel with stable when osm-keystone is promoted
+        await ops_test.juju(
+            "deploy",
+            "--channel=latest/edge",
+            "--trust",
+            "--resource",
+            "keystone-image=opensourcemano/keystone:testing-daily",
+            "osm-keystone",
+            f"{OSM_KEYSTONE_APP_NAME}",
+        )
+
+        await ops_test.model.wait_for_idle(
+            apps=[OSM_KEYSTONE_APP_NAME],
+            status="blocked",
+            raise_on_blocked=False,
+            timeout=1000,
+            wait_for_exact_units=1,
+        )
+
+        # Relate the mysql charm with the osm-keystone charm
+        await ops_test.model.relate(f"{OSM_KEYSTONE_APP_NAME}:db", f"{DATABASE_APP_NAME}:mysql")
+        await ops_test.model.wait_for_idle(
+            apps=[OSM_KEYSTONE_APP_NAME, DATABASE_APP_NAME],
+            status="active",
+            raise_on_blocked=False,
+            timeout=1000,
```

Comment:
Remember that "Not waiting for relations can lead to race-conditions in integration tests", see [this](https://github.com/canonical/mongodb-operator/blob/d49bac6436f7d676be4a4ef6aebf727a245ee815/tests/integration/relation_tests/legacy_relations/test_charm_legacy_relations.py#L186-L201) or Paulo's use of it (I can't find his code for this, maybe you know where it is?). I think Paulo's is nicer since he uses `block_until`. If you don't need to wait, then include a comment as to why [see here for an example](https://github.com/canonical/mongodb-operator/blob/d49bac6436f7d676be4a4ef6aebf727a245ee815/tests/integration/relation_tests/legacy_relations/test_charm_legacy_relations.py#L45)

Reply:
You can also use `block_until` with timeout and a retry, very similar to what we would do with a tenacity retry:
```python

# wait on test
await ops_test.model.block_until(
    lambda: check_application_status(ops_test, "discourse-k8s") == "active",
    timeout=600,
    wait_period=5,
)

def check_application_status(ops_test: OpsTest, app_name: str) -> str:
    """Return the application status for an app name."""
    model_name = ops_test.model.info.name
    proc = subprocess.check_output(f"juju status --model={model_name}".split())
    proc = proc.decode("utf-8")

    statuses = {"active", "maintenance", "waiting", "blocked"}
    for line in proc.splitlines():
        parts = line.split()
        # first line with app name will have application status
        if parts and parts[0] == app_name:
            # NOTE: intersects possible statuses with the list of values:
            # this is done because sometimes version number exists and
            # sometimes it doesn't on juju status output
            find_status = list(statuses & set(parts))
            if find_status:
                return find_status[0]

```